### ISSUE TITLE
Notes on "tuple table"

### ISSUE THREAD

**Issue by da-woods** [cython/cython#6029]
### Describe your issue

One of the things discussed in https://github.com/cython/cython/issues/4425 was to initialize tuples via a "tuple table" in a similar way to the current "string table". My belief is that it's quite difficult to do before https://github.com/cython/cython/pull/5323 is merged (because the global `#define varname modulestate->varname` gets in the way of everything) and so is probably only worthwhile as a follow up.

My notes on a proposed implementation are below (to help remember them later):
* In general tuples can't be put in a table since they depend on other literals being generated. However the vast majority can. Therefore, split tuples into suitable and unsuitable, and use existing code generation for unsuitable.
  * "suitable" can contain:
    * Python constants (`None`, `True`, `False`, `Ellipsis`)
    * integers
    * strings/bytes
    * other suitable tuples
  * "unsuitable" is everything else but might contain `frozenset`, `slices`(?)
  * I propose adding anything that needs multiplication to "unsuitable" because it's a fairly rare complication
* There's two ways we need to be able to lookup objects:
  * Pointer (for things like `None`, `True`, `False`).
  * Offset into modulestate (for most other things)

Therefore my proposal is that the tuple table contains:
* size of tuple (with 0 indicating "stop")
* file index (to keep existing error reporting)
* line no (to keep existing error reporting)
* `N` bits indicating "pointer"/"offset"
* Any padding needed
* `N` unions which can be either a `PyObject*` or a `size_t` depending on pointer/offset

so is really just an array of

```
union TableEntry {
    Py_ssize_t tuple_size;
    int file_line_index;
    size_t mode_selectors;
    size_t offset;
    PyObject *object;
};

**Comment by scoder** [comment]
> Therefore, split tuples into suitable and unsuitable, and use existing code generation for unsuitable.

Don't we do that already with the `is_literal` and `is_partly_literal` flags?

**Comment by da-woods** [comment]
> Don't we do that already with the `is_literal` and `is_partly_literal` flags?

Possibly? I think that `is_partly_literal` is about the multiplication factor.

What I was worried about was if you'd managed to make a tuple that contained another literal initialized at the same time (like `slice` for example), which would work now because things are processed in order, but not if we did tuples before slices. I'm not sure it's possible to get a slice literal into a tuple though, but I don't know that it isn't.

Either way - it's a detail of a rough outline for later

**Comment by TeamSpen210** [comment]
Slices can hold any object whatsoever, `obj[(1,2):3, 4]` works fine. Though I’ve never heard of that being used for anything, probably better to just bail to regular generated code in those cases. It’d make more sense to do tuples after slices, since `array[1:2, 3:4]` would be common in Numpy etc code.

**Comment by scoder** [comment]
> Slices can hold any object whatsoever, `obj[(1,2):3, 4]` works fine.

We're only dealing with immutable constants here. Everything else is handled at runtime rather than module initialisation time.

**Comment by scoder** [comment]
Yes, you can have tuples with slices and slices with tuples. Difficult to map this to a table then.
